source/octf/c/iotrace_plugin.cpp

```diff
@@ -0,0 +1,202 @@
+/*
+ * Copyright(c) 2012-2018 Intel Corporation
+ * SPDX-License-Identifier: BSD-3-Clause-Clear
+ */
+
+#include <octf/c/iotrace_plugin.h>
+
+#include <atomic>
+#include <chrono>
+#include <thread>
+#include <vector>
+#include <octf/trace/IOTracePlugin.h>
+#include <octf/utils/Exception.h>
+#include <octf/utils/Log.h>
+
+using namespace octf;
+
+class IOTracePluginC : public IOTracePlugin {
+private:
+    struct QueueContext {
+        std::atomic_bool traceStoping;
+        std::atomic_uint64_t tracingRefCounter;
+    };
+
+public:
+    IOTracePluginC(octf_iotrace_plugin_context_t context,
+                   const std::string &pluginId,
+                   uint32_t queueCount)
+            : IOTracePlugin(pluginId, queueCount)
+            , m_context(context)
+            , m_queueContext(queueCount) {
+        for (auto &queueContext : m_queueContext) {
+            queueContext.tracingRefCounter = 0;
+            queueContext.traceStoping = true;
+        }
+    }
+
+    virtual ~IOTracePluginC() = default;
+
+    bool startTrace() override {
+        bool result = IOTracePlugin::startTrace();
+
+        if (result) {
+            m_context->tracing_active = true;
+            for (auto &queueContext : m_queueContext) {
+                queueContext.traceStoping = false;
+            }
+        }
+
+        return result;
+    }
+
+    bool stopTrace() override {
+        m_context->tracing_active = false;
+
+        for (auto &queueContext : m_queueContext) {
+            queueContext.traceStoping = true;
+        }
+
+        // Poll for all ongoing traces completion
+        while (isTraceOngoing()) {
+            std::this_thread::sleep_for(std::chrono::milliseconds(10));
+        }
+
+        return IOTracePlugin::stopTrace();
+    }
+
+    void push(uint32_t ioQueueId, const void *trace, size_t size) override {
+        if (ioQueueId > getTraceQueueCount()) {
```

Comment:
What's the expected use case here? I think we should exit here as arguments are invalid.

Reply:
It was intentionally. It applies for cases when someone created the IO trace plugin with one tracing queue, however a program has more then one IO queue. Then the user is able to combine may IO queues into one.